pimd/pim_autorp.c

```diff
@@ -0,0 +1,1146 @@
+	 * inet_pton(PIM_AF, PIM_AUTORP_ANNOUNCEMENT_GRP, &grp);
+	 * if (pim_socket_join(pim->autorp->sock, grp,
+	 *              pim_ifp->primary_address,
+	 *              ifp->ifindex, pim_ifp)) {
+	 *  zlog_err("Failed to join group %pI4 on interface %s",
+	 *          &grp, ifp->name);
+	 *  return errno;
+	 * }
+	 */
+
+	if (PIM_DEBUG_AUTORP)
+		zlog_debug("%s: Joined AutoRP groups on interface %s", __func__,
+			   ifp->name);
+
+	return 0;
+}
+
+static int pim_autorp_leave_groups(struct interface *ifp)
+{
+	struct pim_interface *pim_ifp;
+	struct pim_instance *pim;
+	struct pim_autorp *autorp;
+	pim_addr grp;
+
+	pim_ifp = ifp->info;
+	pim = pim_ifp->pim;
+	autorp = pim->autorp;
+
+	inet_pton(PIM_AF, PIM_AUTORP_DISCOVERY_GRP, &grp);
+	if (pim_socket_leave(autorp->sock, grp, pim_ifp->primary_address,
+			     ifp->ifindex, pim_ifp)) {
+		zlog_err("Failed to leave group %pI4 on interface %s", &grp,
+			 ifp->name);
+		return errno;
+	}
+
+	/* TODO: Leave announcement group for AutoRP mapping agent
+	 * inet_pton(PIM_AF, PIM_AUTORP_ANNOUNCEMENT_GRP, &grp);
+	 * if (pim_socket_leave(pim->autorp->sock, grp,
+	 *              pim_ifp->primary_address,
+	 *              ifp->ifindex, pim_ifp)) {
+	 *  zlog_err("Failed to leave group %pI4 on interface %s",
+	 *           &grp, ifp->name);
+	 *  return errno;
+	 * }
+	 */
+
+	if (PIM_DEBUG_AUTORP)
+		zlog_debug("%s: Left AutoRP groups on interface %s", __func__,
+			   ifp->name);
+
+	return 0;
+}
+
+static int pim_autorp_setup(struct pim_autorp *autorp)
+{
+#if defined(HAVE_IP_PKTINFO)
+	int data;
+	socklen_t data_len = sizeof(data);
+#endif
+
+	struct sockaddr_in autorp_addr = { .sin_family = AF_INET,
+					   .sin_addr = { .s_addr = INADDR_ANY },
+					   .sin_port = htons(PIM_AUTORP_PORT) };
+
+	setsockopt_so_recvbuf(autorp->sock, 1024 * 1024 * 8);
+
+#if defined(HAVE_IP_PKTINFO)
+	/* Linux and Solaris IP_PKTINFO */
+	data = 1;
+	if (setsockopt(autorp->sock, PIM_IPPROTO, IP_PKTINFO, &data, data_len)) {
+		zlog_err("Could not set IP_PKTINFO on socket fd=%d: errno=%d: %s",
+			 autorp->sock, errno, safe_strerror(errno));
+		return errno;
+	}
+#endif
+
+	if (set_nonblocking(autorp->sock) < 0) {
+		zlog_err("Could not set non blocking on socket fd=%d: errno=%d: %s",
+			 autorp->sock, errno, safe_strerror(errno));
+		return errno;
+	}
+
+	if (sockopt_reuseaddr(autorp->sock)) {
+		zlog_err("Could not set reuse addr on socket fd=%d: errno=%d: %s",
+			 autorp->sock, errno, safe_strerror(errno));
+		return errno;
+	}
+
+	if (bind(autorp->sock, (const struct sockaddr *)&autorp_addr,
+		 sizeof(autorp_addr)) < 0) {
+		zlog_err("Could not bind socket: %pSUp, fd=%d, errno=%d, %s",
+			 (union sockunion *)&autorp_addr, autorp->sock, errno,
+			 safe_strerror(errno));
+		return errno;
+	}
+
+	if (PIM_DEBUG_AUTORP)
+		zlog_debug("%s: AutoRP finished setup", __func__);
+
+	return 0;
+}
+
+static int pim_autorp_announcement(struct pim_autorp *autorp, uint8_t rpcnt,
```

Comment:
why have a return?  There is no failure case

Reply:
Why was this resolved?  pim_autorp_announcement always returns true?  How can it fail?

Reply:
This is a stub function where the handling of announcement messages will take place when the AutoRP mapping agent is implemented. So right now it doesn't fail because that handling is not implemented but it will have failure cases in the future.